### DISCUSSION TITLE
please explain :)

### DISCUSSION THREAD

**Discussion by istvan-b01** [spakin/SimpInkScr#94] [General]
I know that this is not stackoverflaw but this is now beyond me as all my struggles were likely to originate from this point
if a shape is created at certain place it does not append to a layer despite that the shape is created, while if it;s made at another place than it does.
(extended box drawing character generation on canvas 504x1008)
please see the code below, the append inside of the condition that does the shenanigan...

```
import sys
style(stroke_width=1, fill='rgb(0,0,0)')

PUA = 58367
SIZE = 1008; HEIGHT = SIZE; WIDTH = HEIGHT >> 1

vUnit = HEIGHT // 12; hUnit = WIDTH // 8

thinTop = [(3 * hUnit, 0 * vUnit), (5 * hUnit, 7 * vUnit)]
thinRight = [(3 * hUnit, 5 * vUnit), (8 * hUnit, 7 * vUnit)]
thinBottom = [(3 * hUnit, 5 * vUnit), (5 * hUnit, 12 * vUnit)]
thinLeft = [(0 * hUnit, 5 * vUnit), (5 * hUnit, 7 * vUnit)]

thickTop = [(2 * hUnit, 0 * vUnit), (6 * hUnit, 8 * vUnit)]
thickRight = [(2 * hUnit, 4 * vUnit), (8 * hUnit, 8 * vUnit)]
thickBottom = [(2 * hUnit, 4 * vUnit), (6 * hUnit, 12 * vUnit)]
thickLeft = [(0 * hUnit, 4 * vUnit), (6 * hUnit, 8 * vUnit)]

bars = [[thinTop, thinRight, thinBottom, thinLeft], [thickTop, thickRight, thickBottom, thickLeft]]

for glyph in range(255, -1, -1):
    arrOuter = []; arrInner = []
    sector = glyph
    for quadrant in range(0, 4, 1):
        barType = sector & 3
        match barType:
            case 1 | 2:
                arrOuter.append(rect(bars[barType - 1][quadrant][0], bars[barType - 1][quadrant][1]).to_path())
            case 3:
                arrOuter.append(rect(bars[1][quadrant][0], bars[1][quadrant][1]).to_path())
                arrInner.append(rect(bars[0][quadrant][0], bars[0][quadrant][1]).to_path())
        sector >>= 2
    name = hex(PUA).split('x')[-1]; character = layer('U_' + name.upper())
    if len(arrOuter) != 0:
        shape = arrOuter[0] if len(arrOuter) == 1 else apply_path_operation('union', arrOuter)[0]
        if len(arrInner) != 0:
            extract = arrInner[0] if len(arrInner) == 1 else apply_path_operation('union', arrInner)[0]
            shape = apply_path_operation('difference', [shape, extract])[0]
            #character.append(shape)
            sys.exit()
        character.append(shape)
    PUA -= 1
```

====================================================================================
if the glyph loop starts from 0 then it does append the shapes in cases where no extraction is not invoked

please tell me what's happening :)

**Comment by istvan-b01** [comment]
if the character.append(shape) is uncommented then the shape goes into thin air and results in an empty layer.
linux mint, inkscape 1.3

**Comment by istvan-b01** [comment]
well, new ideas come after question has been asked.
the layer has to be created right before appending, then it works. I don't understand the reason, but that solves it.

**Comment by spakin** [comment]
I can reproduce the problem.  The following simpler variant of your code also sends the shape into thin air:
```Python
my_layer = layer('My Layer')
r1 = rect((0, 0), (300, 300), fill='darkblue').to_path()
r2 = rect((100, 100), (200, 200), fill='orange').to_path()
shape = apply_path_operation('difference', [r1, r2])[0]
my_layer.append(shape)
```
The shape appears correct if it's not moved from the default layer:
```Python
my_layer = layer('My Layer')
r1 = rect((0, 0), (300, 300), fill='darkblue').to_path()
r2 = rect((100, 100), (200, 200), fill='orange').to_path()
shape = apply_path_operation('difference', [r1, r2])[0]
```
Also, shapes created without path operations can move successfully to the new layer:
```Python
my_layer = layer('My Layer')
r1 = rect((0, 0), (300, 300), fill='darkblue').to_path()
r2 = rect((100, 100), (200, 200), fill='orange').to_path()
my_layer.extend([r1, r2])
```

Therefore, there must be some odd interaction between path operations and layers.  I'm looking into this but am finding it tricky to identify the source of the problem.

**Reply by spakin** [reply]
I think I figured it out!  Please download the latest commit of Simple Inkscape Scripting from GitHub and let me know if that fixes the problem—and doesn't break anything else.

**Reply by istvan-b01** [reply]
thanks for attending the issue. I'm glad that accidentally I was able to contribute a bit in the improving of the project.
I general, I'm still new to git (in terms of actively using it) so I couldn't really find the place to where I could download it.

in the meantime and probably the final time I got another problem that is crucial for finally finishing what I wanted to do.
That is about saving the file.
First is there a way to add a path to the filename in order to save the document in a specific location? When I tried to do that, the path was showing correctly but was saying that the file does not exist (e.g.: path/filename.svg) where the directory varies throughout the course of the script.

So the task is, that I'm generating over 2000 glyphs (old micro computer's character sets) to different fonts, and in order to be able to import to fontforge, each glyph has to be in a separate file.
when a glyph is saved then it should cease to exist keeping the name for the next one.

When I skipped the 'path' from the file name the code run, but no files were saved as far as I know, the glyphs have been generated all crowded up , despite I used the 'del' keyword on every one of them.

So what am I missing in the interpretation?
I'm new to python, pretty much to inkscape too, let alone to both of the in the same time so I might miss something obvious.

Also, could you please write a short code that covers pretty much what I need to achieve.
I'm not being lazy, but the description of the 'save_file' is pretty clear and I still didn't manage to get it work.

Thanks again.

ps: please include a link for the commit you mentioned

**Reply by spakin** [reply]
> thanks for attending the issue. I'm glad that accidentally I was able to contribute a bit in the improving of the project.
> I general, I'm still new to git (in terms of actively using it) so I couldn't really find the place to where I could download it.

First, uninstall the current version of Simple Inkscape Scripting (from the Extensions Manager if you installed it that way; otherwise remove the `simple_inkscape_scripting` directory).

Then, if you don't want to deal with Git, the easiest alternative is to go to https://github.com/spakin/SimpInkScr, click the green Code button, and select Download ZIP.  Unzip that into `~/.config/inkscape/extensions/` or wherever your Inkscape extensions live.

Alternatively, if you _do_ want to work with Git, you'll need to clone the GitHub repository locally.  The green Code button presents the URL to use for either SSH downloads (requires first giving GitHub your SSH public key) or HTTPS downloads or a command line using GitHub's command-line tool, `gh` (which I've not used myself).  For the first two, pass the URL to `git clone` with a command like the following:

```bash
git clone https://github.com/spakin/SimpInkScr.git ~/.config/inkscape/extensions/simple_inkscape_scripting
```

Any time thereafter you wish to download the latest version of Simple Inkscape Scripting, simply run
```bash
git pull
```
from within the `simple_inkscape_scripting` directory.

> in the meantime and probably the final time I got another problem that is crucial for finally finishing what I wanted to do.
> That is about saving the file.
> First is there a way to add a path to the filename in order to save the document in a specific location? When I tried to do that, the path was showing correctly but was saying that the file does not exist (e.g.: path/filename.svg) where the directory varies throughout the course of the script.

`save_file` should work as expected if you give it an absolute path (barring `/tmp`, which is silently renamed if you're running a Snap install of Inkscape).  Relative paths are relative to the extensions directory, not the script directory.

In a Simple Inkscape Scripting script, [`script_filename`](https://github.com/spakin/SimpInkScr/wiki/Other-features#source-filename) returns the script's absolute filename.  Python's [`os.path.dirname`](https://docs.python.org/3/library/os.path.html#os.path.dirname) can extract the directory name from that in case you want to generate a filename relative to the script directory.

> Also, could you please write a short code that covers pretty much what I need to achieve.
> I'm not being lazy, but the description of the 'save_file' is pretty clear and I still didn't manage to get it work.

Something as simple as
```Python
star(7, (100, 100), (80, 30), fill='yellow', stroke_width=2)
save_file('/home/pakin/example.svg')
```
works for me.

**Reply by istvan-b01** [reply]
yep, that was my bad, if I use absolute path for the file name then they get saved beautifully, thanks, and sorry for the effortless attempt.
The last problem is still outstanding , which is how can I destroy a layer or shape after saving them as the 'del' does not really do the job therefore at the moment they are still desperately clinging to their existence till the end of the script which, cruel or not, is not in my interest as I need one layer/shape per file.

I guess to keep everyone happy, making them invisible in the script may work as well, although fontforge may think otherwise.
based on what you showed me regarding renaming the objects I tried

layer0.svg_set('inkscape:visible', False)

but this didn't do much. Calling the attribute 'visible' was just a wild guess

**Reply by istvan-b01** [reply]
Sorry, I did it again, a couple of google search on inkex delete, yielded me the clear() method in the turtle module and gave me everything I needed.

Thank you very much for assisting me along, I learnt a great deal in the meantime and gave me enough to know where to look in the future if I need help. But in case I won't find it I'll still going to ask :)

**Comment by spakin** [comment]
All Simple Inkscape Scripting shapes support a [`remove`](https://github.com/spakin/SimpInkScr/wiki/Modifying-existing-objects#removing-an-existing-object) method:
```Python
c = circle((50, 50), 50, fill='red')
save_file('/home/me/circle.svg')
c.remove()
r = rect((100,0), (200, 100), fill='limegreen')
save_file('/home/me/square.svg')
```

**Reply by istvan-b01** [reply]
Thanks again, for addressing. In this case it did the job, but it didn't fit my need as somehow left a thin margin on the side, although this might have been some type of oversight on my side.
Putting the shape in a layer then clearing the layer did the trick. However, following the old paradigm, if a program works then you need to change it, I decided that I'd like to create the negative of each glyph as well and run into trouble straight away.

If you could please look at the code to see what am I missing (sorry for coding style, probably I'm breaking all python's coding standards)
In the main part glyph is already existing and it's either an empty shape /1 rectangle / multiple rectangles combine by  union.
It works well, and it saves it with the save function.
However afterward I wanted to generate the inverse shape and I thing I still don't know what is going on. What is a shape and what isn't (if glyph is a rectangle it behaves differently to if it's the result of a path operation.
So, if I create the inverse it does not do the 'difference'. Thinking that the full canvas size BLOCK is still on the canvnas and covering the result I tried to remove both, but in case of glyph0 it complains for not having a delete function

    def save_glyph(directory, unicode, GLYPH):
    name = 'uni' + ('0000' + hex(unicode)[2:]).upper()[-4:]
    layer0 = layer('glyph', GLYPH)
    save_file(BASE_DIR + directory + name + '.svg')
    layer0.clear()

        save_glyph(Directory_Name[font0], ascii_code, glyph)
        BLOCK = rect((0, 0), (WIDTH - 1, HEIGHT - 1)).to_path();
        glyph0 = BLOCK if bits == 0 else apply_path_operation('difference', [BLOCK, glyph])[0]
        save_glyph(Directory_Name[font0], ascii_code + 57344, glyph0)    # E000


Btw, is there a something like shape.inverse() method that turn a path into it's inverse regarding to the canvas, like in photoshop I think CTRL-D or something similar?

Thanks again for the continuous help!

**Comment by istvan-b01** [comment]
Well, as before, I got the solution again by thinking and coming up with the obvious. Turning glyph into a path again and then using the path operator did the trick.Anyway, thank you very much for following through and giving good ideas about the potential of Simple Inkscape Script.
    On Wednesday, 31 May 2023 at 02:14:44 BST, Scott Pakin ***@***.***> wrote:  
 
 


All Simple Inkscape Scripting shapes support a remove method:
c = circle((50, 50), 50, fill='red')
save_file('/home/me/circle.svg')
c.remove()
r = rect((100,0), (200, 100), fill='limegreen')
save_file('/home/me/square.svg')
